adapters/interlocked_linux.c

```diff
@@ -1,4 +1,296 @@
+    /*Codes_SRS_INTERLOCKED_LINUX_43_004: [ interlocked_and shall return the initial value of *destination. ]*/
+    return atomic_fetch_and(destination, value);
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, int16_t, interlocked_and_16, volatile_atomic int16_t*, destination, int16_t, value)
+{
+    /*Codes_SRS_INTERLOCKED_LINUX_43_005: [ interlocked_and_16 shall call atomic_fetch_and with destination as object and value as operand. ]*/
+    /*Codes_SRS_INTERLOCKED_LINUX_43_006: [ interlocked_and_16 shall return the initial value of *destination. ]*/
+    return atomic_fetch_and(destination, value);
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, int64_t, interlocked_and_64, volatile_atomic int64_t*, destination, int64_t, value)
+{
+    /*Codes_SRS_INTERLOCKED_LINUX_43_007: [ interlocked_and_64 shall call atomic_fetch_and with destination as object and value as operand. ]*/
+    /*Codes_SRS_INTERLOCKED_LINUX_43_008: [ interlocked_and_64 shall return the initial value of *destination. ]*/
+    return atomic_fetch_and(destination, value);
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, int8_t, interlocked_and_8, volatile_atomic int8_t*, destination, int8_t, value)
+{
+    /*Codes_SRS_INTERLOCKED_LINUX_43_009: [ interlocked_and_8 shall call atomic_fetch_and with destination as object and value as operand. ]*/
+    /*Codes_SRS_INTERLOCKED_LINUX_43_010: [ interlocked_and_8 shall return the initial value of *destination. ]*/
+    return atomic_fetch_and(destination, value);
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, int32_t, interlocked_compare_exchange, volatile_atomic int32_t*, destination, int32_t, exchange, int32_t, comperand)
+{
+    while(true)
+    {
+        int32_t comperand_copy = comperand;
+        /*Codes_SRS_INTERLOCKED_LINUX_43_011: [ interlocked_compare_exchange shall call atomic_compare_exchange_strong with destination as object, &comperand as expected and exchange as desired. ]*/
+        /*Codes_SRS_INTERLOCKED_LINUX_43_012: [ interlocked_compare_exchange shall return a value equal to comperand if and only if atomic_compare_exchange_strong returns true. ]*/
+        if(atomic_compare_exchange_strong(destination, &comperand_copy, exchange))
+        {
+            return comperand;
+        }
+        else
+        {
+            /*Codes_SRS_INTERLOCKED_LINUX_43_066: [ interlocked_compare_exchange shall call atomic_load if atomic_compare_exchange_strong returns false. ]*/
+            int32_t loaded_value = atomic_load(destination);
+            if(loaded_value != comperand)
+            {
+                /*Codes_SRS_INTERLOCKED_LINUX_43_067: [ interlocked_compare_exchange shall return the value returned by atomic_load if the returned value is not equal to comperand. ]*/
+                return loaded_value;
+            }
+        }
+        /*Codes_SRS_INTERLOCKED_LINUX_43_068: [ interlocked_compare_exchange shall repeat the call to atomic_compare_exchange_strong if the value returned by atomic_load is equal to comperand. ]*/
+    }
```

Comment:
Sorry, we didn't get a chance to discuss this yesterday, but after you mentioned that aces modifies the second arg ("comparand"), I said we actually can change this.

Spec says:
"Atomically, compares the value pointed to by object for equality
with that in expected, and if true, replaces the value pointed to by object with
desired, and *if false, updates the value in expected with the value pointed to by
object.* "

So really we just need to return comparand_copy always, much simpler :)

I think all we need is this (obviously, you can remove most of the specs too to simplify)

```
int32_t comperand_copy = comperand;
(void)atomic_compare_exchange_strong(destination, &comperand_copy, exchange);
return comparand_copy;
```

comparand_copy is the same as comparand if atomic_compare_exchange returns true.

comparand_copy is the original/current value in destination if atomic_compare_exchange returns false.

Reply:
Since comperand is passed by value to interlocked_compare_exchange, is it necessary to make a copy of it?

---
In reply to: [433989674](https://github.com/Azure/azure-c-util/pull/46#discussion_r433989674) [](ancestors = 433989674)

Reply:
It's not necessary but I think it's cleaner to avoid modifying arguments (IMO all func args should be const, but declaring them as const is overly verbose).

Making a copy doesn't hurt as the compiler will optimize it away anyway.

---
In reply to: [433992371](https://github.com/Azure/azure-c-util/pull/46#discussion_r433992371) [](ancestors = 433992371,433989674)